cmd/dockerd/daemon.go

```diff
@@ -244,6 +244,17 @@ func (cli *daemonCLI) start(ctx context.Context) (err error) {
 
 	pluginStore := plugin.NewStore()
 
+	// Register the CDI driver before the daemon starts, as it might try to restore containers that depend on the CDI driver.
+	// Note that CDI is not inherently linux-specific, there are some linux-specific assumptions / implementations in the code that
+	// queries the properties of device on the host as well as performs the injection of device nodes and their access permissions into the OCI spec.
+	//
+	// In order to lift this restriction the following would have to be addressed:
+	// - Support needs to be added to the cdi package for injecting Windows devices: https://tags.cncf.io/container-device-interface/issues/28
+	// - The DeviceRequests API must be extended to non-linux platforms.
+	if runtime.GOOS == "linux" && cli.Config.Features["cdi"] {
+		daemon.RegisterCDIDriver(cli.Config.CDISpecDirs...)
+	}
+
```

Comment:
Change looks good to me; we should probably look at rolling this into `daemon.NewDaemon`, as there's some validation and setup steps being performed in that function as well (and _possibly_ adjusting of config options), so possibly this should go after that; https://github.com/moby/moby/blob/b87aecc50adb5af7c79b9fcffde0fed79735b115/daemon/daemon.go#L763-L783